File: src/dynamic_programming/cpp/minimum_insertions_for_palindrome/src/minimum_insertions_for_palindrome.cpp
#include "minimum_insertions_for_palindrome.h"
#include <climits>
#include <functional>

int minimum_insertions_for_palindrome_basic(const std::string &s) {
  int n = s.length();
  if (n == 0) {
    return 0;
  }

  std::function<int(const std::string &, int, int)> min_insertions_helper;
  min_insertions_helper = [&min_insertions_helper](const std::string &s, int l,
                                                   int h) {
    if (l > h) {
      return INT_MAX;
    }

    if (l == h) {
      return 0;
    }

    if (l == h - 1) {
      return s[l] == s[h] ? 0 : 1;
    }

    if (s[l] == s[h]) {
      return min_insertions_helper(s, l + 1, h - 1);
    } else {
      return 1 + std::min(min_insertions_helper(s, l, h - 1),
                          min_insertions_helper(s, l + 1, h));
    }
  };

  return min_insertions_helper(s, 0, n - 1);
}

int minimum_insertions_for_palindrome_memo(const std::string &s) {
  int n = s.length();
  if (n == 0) {
    return 0;
  }

  std::vector<std::vector<int>> memo(n, std::vector<int>(n, -1));
  std::function<int(const std::string &, int, int)> min_insertions_helper;
  min_insertions_helper = [&min_insertions_helper, &memo](const std::string &s,
                                                          int l, int h) {
    if (l > h) {
      return INT_MAX;
    }

    if (l == h) {
      return 0;
    }

    if (l == h - 1) {
      return s[l] == s[h] ? 0 : 1;
    }

    if (memo[l][h] != -1) {
      return memo[l][h];
    }

    if (s[l] == s[h]) {
      memo[l][h] = min_insertions_helper(s, l + 1, h - 1);
    } else {
      memo[l][h] = 1 + std::min(min_insertions_helper(s, l, h - 1),
                                min_insertions_helper(s, l + 1, h));
    }

    return memo[l][h];
  };

  return min_insertions_helper(s, 0, n - 1);
}

int minimum_insertions_for_palindrome_tab(const std::string &s) {
  int n = s.length();
  if (n == 0) {
    return 0;
  }

  std::vector<std::vector<int>> table(n, std::vector<int>(n, 0));
  for (int i = 0; i < n; i++) {
    table[i][i] = 0;
  }

  for (int l = 2; l <= n; l++) {
    for (int i = 0; i < n - l + 1; i++) {
      int j = i + l - 1;
      if (l == 2) {
        table[i][j] = s[i] == s[j] ? 0 : 1;
      } else {
        if (s[i] == s[j]) {
          table[i][j] = table[i + 1][j - 1];
        } else {
          table[i][j] = 1 + std::min(table[i][j - 1], table[i + 1][j]);
        }
      }
    }
  }

  return table[0][n - 1];
}

File: src/dynamic_programming/python/minimum_insertions_for_palindrome/src/minimum_insertions_for_palindrome.py
from typing import List


def minimum_insertions_for_palindrome_basic(s: str) -> int:
    """
    Find minimum insertions needed to make a string a palindrome.

    Uses basic recursion without memoization.

    Args:
        s: The input string.

    Returns:
        Minimum number of character insertions needed.
    """
    n = len(s)
    if n == 0:
        return 0

    def min_insertions_helper(s: str, l: int, h: int) -> int:
        if l > h:
            return float("inf")

        if l == h:
            return 0

        if l == h - 1:
            return 0 if s[l] == s[h] else 1

        if s[l] == s[h]:
            return min_insertions_helper(s, l + 1, h - 1)
        else:
            return 1 + min(
                min_insertions_helper(s, l, h - 1), min_insertions_helper(s, l + 1, h)
            )

    return min_insertions_helper(s, 0, n - 1)


def minimum_insertions_for_palindrome_memo(s: str) -> int:
    """
    Find minimum insertions needed to make a string a palindrome.

    Uses memoization for efficient computation.

    Args:
        s: The input string.

    Returns:
        Minimum number of character insertions needed.
    """
    n = len(s)
    if n == 0:
        return 0

    def min_insertions_helper(s: str, l: int, h: int, memo: dict) -> int:
        if l > h:
            return float("inf")

        if l == h:
            return 0

        if l == h - 1:
            return 0 if s[l] == s[h] else 1

        key = (l, h)
        if key in memo:
            return memo[key]

        if s[l] == s[h]:
            memo[key] = min_insertions_helper(s, l + 1, h - 1, memo)
        else:
            memo[key] = 1 + min(
                min_insertions_helper(s, l, h - 1, memo),
                min_insertions_helper(s, l + 1, h, memo),
            )
        return memo[key]

    return min_insertions_helper(s, 0, n - 1, {})


def minimum_insertions_for_palindrome_tab(s: str) -> int:
    """
    Find minimum insertions needed to make a string a palindrome.

    Uses tabulation (bottom-up DP) for efficient computation.

    Args:
        s: The input string.

    Returns:
        Minimum number of character insertions needed.
    """
    n = len(s)
    if n == 0:
        return 0

    table: List[List[int]] = [[0 for _ in range(n)] for _ in range(n)]

    for i in range(n - 1, -1, -1):
        for j in range(i + 1, n):
            if s[i] == s[j]:
                table[i][j] = table[i + 1][j - 1]
            else:
                table[i][j] = 1 + min(table[i + 1][j], table[i][j - 1])

    return table[0][-1]
